File: sdk/typescript/src/namespaces/orchestration_canvas.ts
/**
 * Orchestration Canvas Namespace API
 *
 * Provides endpoints for the Orchestration Canvas (Stage 4 of the
 * Idea-to-Execution Pipeline): canvas CRUD, node/edge management,
 * React Flow export, and pipeline execution.
 */

import type { AragoraClient } from '../client';

/** An orchestration node on the canvas */
export interface OrchestrationNode {
  id: string;
  label: string;
  orchestration_type: string;
  position?: { x: number; y: number };
  data?: Record<string, unknown>;
}

/** An edge connecting two orchestration nodes */
export interface OrchestrationEdge {
  id: string;
  source_id: string;
  target_id: string;
  edge_type?: string;
  label?: string;
  data?: Record<string, unknown>;
}

/** An orchestration canvas document */
export interface OrchestrationCanvas {
  id: string;
  name: string;
  description?: string;
  owner_id?: string;
  workspace_id?: string;
  source_canvas_id?: string;
  metadata?: Record<string, unknown>;
  nodes?: OrchestrationNode[];
  edges?: OrchestrationEdge[];
  created_at: string;
  updated_at: string;
}

/** Result of executing the pipeline */
export interface ExecutionResult {
  execution_id: string;
  canvas_id: string;
  stage: string;
  nodes_count: number;
  edges_count: number;
  metadata: Record<string, unknown>;
  status: string;
}

/**
 * Orchestration Canvas namespace (Stage 4 of the
 * Idea-to-Execution Pipeline).
 */
export class OrchestrationCanvasNamespace {
  constructor(private client: AragoraClient) {}

  // -------------------------------------------------------------------
  // Canvas CRUD
  // -------------------------------------------------------------------

  /** List orchestration canvases with optional filtering. */
  async listCanvases(options?: {
    workspace_id?: string;
    owner_id?: string;
    source_canvas_id?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ canvases: OrchestrationCanvas[] }> {
    return this.client.request<{ canvases: OrchestrationCanvas[] }>(
      'GET',
      '/api/v1/orchestration/canvas',
      { params: options },
    );
  }

  /** Create a new orchestration canvas. */
  async createCanvas(
    name: string,
    description: string = '',
    metadata?: Record<string, unknown>,
  ): Promise<OrchestrationCanvas> {
    const body: Record<string, unknown> = { name, description };
    if (metadata !== undefined) {
      body.metadata = metadata;
    }
    return this.client.request<OrchestrationCanvas>(
      'POST',
      '/api/v1/orchestration/canvas',
      { body },
    );
  }

  /** Get a canvas by ID, including its nodes and edges. */
  async getCanvas(canvasId: string): Promise<OrchestrationCanvas> {
    return this.client.request<OrchestrationCanvas>(
      'GET',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}`,
    );
  }

  /** Update canvas metadata. */
  async updateCanvas(
    canvasId: string,
    updates: {
      name?: string;
      description?: string;
      metadata?: Record<string, unknown>;
    },
  ): Promise<OrchestrationCanvas> {
    return this.client.request<OrchestrationCanvas>(
      'PUT',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}`,
      { body: updates },
    );
  }

  /** Delete a canvas and all its nodes and edges. */
  async deleteCanvas(canvasId: string): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Node CRUD
  // -------------------------------------------------------------------

  /** Add a node to a canvas. */
  async addNode(
    canvasId: string,
    label: string,
    orchestrationType: string = 'agent_task',
    position?: { x: number; y: number },
    data?: Record<string, unknown>,
  ): Promise<OrchestrationNode> {
    const body: Record<string, unknown> = {
      label,
      orchestration_type: orchestrationType,
    };
    if (position !== undefined) {
      body.position = position;
    }
    if (data !== undefined) {
      body.data = data;
    }
    return this.client.request<OrchestrationNode>(
      'POST',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/nodes`,
      { body },
    );
  }

  /** Update a node on a canvas. */
  async updateNode(
    canvasId: string,
    nodeId: string,
    updates: {
      label?: string;
      position?: { x: number; y: number };
      data?: Record<string, unknown>;
    },
  ): Promise<OrchestrationNode> {
    return this.client.request<OrchestrationNode>(
      'PUT',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/nodes/${encodeURIComponent(nodeId)}`,
      { body: updates },
    );
  }

  /** Delete a node from a canvas. */
  async deleteNode(
    canvasId: string,
    nodeId: string,
  ): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/nodes/${encodeURIComponent(nodeId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Edge CRUD
  // -------------------------------------------------------------------

  /** Add an edge connecting two nodes on a canvas. */
  async addEdge(
    canvasId: string,
    sourceId: string,
    targetId: string,
    edgeType: string = 'default',
    label: string = '',
    data?: Record<string, unknown>,
  ): Promise<OrchestrationEdge> {
    const body: Record<string, unknown> = {
      source_id: sourceId,
      target_id: targetId,
      edge_type: edgeType,
      label,
    };
    if (data !== undefined) {
      body.data = data;
    }
    return this.client.request<OrchestrationEdge>(
      'POST',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/edges`,
      { body },
    );
  }

  /** Delete an edge from a canvas. */
  async deleteEdge(
    canvasId: string,
    edgeId: string,
  ): Promise<{ success: boolean }> {
    return this.client.request<{ success: boolean }>(
      'DELETE',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/edges/${encodeURIComponent(edgeId)}`,
    );
  }

  // -------------------------------------------------------------------
  // Export & Execute
  // -------------------------------------------------------------------

  /** Export a canvas as React Flow JSON (nodes + edges arrays). */
  async exportCanvas(canvasId: string): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'GET',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/export`,
    );
  }

  /** Execute the orchestration pipeline defined by this canvas. */
  async executePipeline(canvasId: string): Promise<ExecutionResult> {
    return this.client.request<ExecutionResult>(
      'POST',
      `/api/v1/orchestration/canvas/${encodeURIComponent(canvasId)}/execute`,
      { body: {} },
    );
  }
}

File: sdk/python/aragora_sdk/namespaces/orchestration_canvas.py
"""
Orchestration Canvas Namespace API

Provides methods for the Orchestration Canvas (Stage 4 of the Idea-to-Execution Pipeline):
- Canvas CRUD (list, create, get, update, delete)
- Node CRUD (add, update, delete)
- Edge CRUD (add, delete)
- Export to React Flow JSON
- Execute pipeline
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class OrchestrationCanvasAPI:
    """Synchronous Orchestration Canvas API."""

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # -------------------------------------------------------------------
    # Canvas CRUD
    # -------------------------------------------------------------------

    def list_canvases(
        self,
        workspace_id: str | None = None,
        owner_id: str | None = None,
        source_canvas_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List orchestration canvases with optional filtering."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        if owner_id is not None:
            params["owner_id"] = owner_id
        if source_canvas_id is not None:
            params["source_canvas_id"] = source_canvas_id
        return self._client.request("GET", "/api/v1/orchestration/canvas", params=params)

    def create_canvas(
        self,
        name: str,
        description: str = "",
        source_canvas_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a new orchestration canvas."""
        body: dict[str, Any] = {"name": name, "description": description}
        if source_canvas_id is not None:
            body["source_canvas_id"] = source_canvas_id
        if metadata is not None:
            body["metadata"] = metadata
        return self._client.request("POST", "/api/v1/orchestration/canvas", json=body)

    def get_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Get a canvas by ID, including its nodes and edges."""
        return self._client.request("GET", f"/api/v1/orchestration/canvas/{canvas_id}")

    def update_canvas(
        self,
        canvas_id: str,
        name: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update canvas metadata."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if metadata is not None:
            body["metadata"] = metadata
        return self._client.request("PUT", f"/api/v1/orchestration/canvas/{canvas_id}", json=body)

    def delete_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Delete a canvas and all its nodes and edges."""
        return self._client.request("DELETE", f"/api/v1/orchestration/canvas/{canvas_id}")

    # -------------------------------------------------------------------
    # Node CRUD
    # -------------------------------------------------------------------

    def add_node(
        self,
        canvas_id: str,
        label: str,
        orchestration_type: str = "agent_task",
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add a node to a canvas."""
        body: dict[str, Any] = {"label": label, "orchestration_type": orchestration_type}
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return self._client.request(
            "POST", f"/api/v1/orchestration/canvas/{canvas_id}/nodes", json=body
        )

    def update_node(
        self,
        canvas_id: str,
        node_id: str,
        label: str | None = None,
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update a node on a canvas."""
        body: dict[str, Any] = {}
        if label is not None:
            body["label"] = label
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return self._client.request(
            "PUT", f"/api/v1/orchestration/canvas/{canvas_id}/nodes/{node_id}", json=body
        )

    def delete_node(self, canvas_id: str, node_id: str) -> dict[str, Any]:
        """Delete a node from a canvas."""
        return self._client.request(
            "DELETE", f"/api/v1/orchestration/canvas/{canvas_id}/nodes/{node_id}"
        )

    # -------------------------------------------------------------------
    # Edge CRUD
    # -------------------------------------------------------------------

    def add_edge(
        self,
        canvas_id: str,
        source_id: str,
        target_id: str,
        edge_type: str = "default",
        label: str = "",
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add an edge connecting two nodes on a canvas."""
        body: dict[str, Any] = {
            "source_id": source_id,
            "target_id": target_id,
            "edge_type": edge_type,
            "label": label,
        }
        if data is not None:
            body["data"] = data
        return self._client.request(
            "POST", f"/api/v1/orchestration/canvas/{canvas_id}/edges", json=body
        )

    def delete_edge(self, canvas_id: str, edge_id: str) -> dict[str, Any]:
        """Delete an edge from a canvas."""
        return self._client.request(
            "DELETE", f"/api/v1/orchestration/canvas/{canvas_id}/edges/{edge_id}"
        )

    # -------------------------------------------------------------------
    # Export & Execute
    # -------------------------------------------------------------------

    def export_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Export a canvas as React Flow JSON."""
        return self._client.request("GET", f"/api/v1/orchestration/canvas/{canvas_id}/export")

    def execute_pipeline(self, canvas_id: str) -> dict[str, Any]:
        """Execute the orchestration pipeline defined by this canvas."""
        return self._client.request(
            "POST",
            f"/api/v1/orchestration/canvas/{canvas_id}/execute",
            json={},
        )


class AsyncOrchestrationCanvasAPI:
    """Asynchronous Orchestration Canvas API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # -------------------------------------------------------------------
    # Canvas CRUD
    # -------------------------------------------------------------------

    async def list_canvases(
        self,
        workspace_id: str | None = None,
        owner_id: str | None = None,
        source_canvas_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List orchestration canvases with optional filtering."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        if owner_id is not None:
            params["owner_id"] = owner_id
        if source_canvas_id is not None:
            params["source_canvas_id"] = source_canvas_id
        return await self._client.request("GET", "/api/v1/orchestration/canvas", params=params)

    async def create_canvas(
        self,
        name: str,
        description: str = "",
        source_canvas_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a new orchestration canvas."""
        body: dict[str, Any] = {"name": name, "description": description}
        if source_canvas_id is not None:
            body["source_canvas_id"] = source_canvas_id
        if metadata is not None:
            body["metadata"] = metadata
        return await self._client.request("POST", "/api/v1/orchestration/canvas", json=body)

    async def get_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Get a canvas by ID, including its nodes and edges."""
        return await self._client.request("GET", f"/api/v1/orchestration/canvas/{canvas_id}")

    async def update_canvas(
        self,
        canvas_id: str,
        name: str | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update canvas metadata."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if metadata is not None:
            body["metadata"] = metadata
        return await self._client.request(
            "PUT", f"/api/v1/orchestration/canvas/{canvas_id}", json=body
        )

    async def delete_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Delete a canvas and all its nodes and edges."""
        return await self._client.request("DELETE", f"/api/v1/orchestration/canvas/{canvas_id}")

    # -------------------------------------------------------------------
    # Node CRUD
    # -------------------------------------------------------------------

    async def add_node(
        self,
        canvas_id: str,
        label: str,
        orchestration_type: str = "agent_task",
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add a node to a canvas."""
        body: dict[str, Any] = {"label": label, "orchestration_type": orchestration_type}
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return await self._client.request(
            "POST", f"/api/v1/orchestration/canvas/{canvas_id}/nodes", json=body
        )

    async def update_node(
        self,
        canvas_id: str,
        node_id: str,
        label: str | None = None,
        position: dict[str, Any] | None = None,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update a node on a canvas."""
        body: dict[str, Any] = {}
        if label is not None:
            body["label"] = label
        if position is not None:
            body["position"] = position
        if data is not None:
            body["data"] = data
        return await self._client.request(
            "PUT", f"/api/v1/orchestration/canvas/{canvas_id}/nodes/{node_id}", json=body
        )

    async def delete_node(self, canvas_id: str, node_id: str) -> dict[str, Any]:
        """Delete a node from a canvas."""
        return await self._client.request(
            "DELETE", f"/api/v1/orchestration/canvas/{canvas_id}/nodes/{node_id}"
        )

    # -------------------------------------------------------------------
    # Edge CRUD
    # -------------------------------------------------------------------

    async def add_edge(
        self,
        canvas_id: str,
        source_id: str,
        target_id: str,
        edge_type: str = "default",
        label: str = "",
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Add an edge connecting two nodes on a canvas."""
        body: dict[str, Any] = {
            "source_id": source_id,
            "target_id": target_id,
            "edge_type": edge_type,
            "label": label,
        }
        if data is not None:
            body["data"] = data
        return await self._client.request(
            "POST", f"/api/v1/orchestration/canvas/{canvas_id}/edges", json=body
        )

    async def delete_edge(self, canvas_id: str, edge_id: str) -> dict[str, Any]:
        """Delete an edge from a canvas."""
        return await self._client.request(
            "DELETE", f"/api/v1/orchestration/canvas/{canvas_id}/edges/{edge_id}"
        )

    # -------------------------------------------------------------------
    # Export & Execute
    # -------------------------------------------------------------------

    async def export_canvas(self, canvas_id: str) -> dict[str, Any]:
        """Export a canvas as React Flow JSON."""
        return await self._client.request("GET", f"/api/v1/orchestration/canvas/{canvas_id}/export")

    async def execute_pipeline(self, canvas_id: str) -> dict[str, Any]:
        """Execute the orchestration pipeline defined by this canvas."""
        return await self._client.request(
            "POST",
            f"/api/v1/orchestration/canvas/{canvas_id}/execute",
            json={},
        )
